diff -up nss-pam-ldapd-0.7.5/nslcd/passwd.c.growgecos nss-pam-ldapd-0.7.5/nslcd/passwd.c
--- nss-pam-ldapd-0.7.5/nslcd/passwd.c.growgecos	2015-11-13 10:40:29.449555225 +0100
+++ nss-pam-ldapd-0.7.5/nslcd/passwd.c	2015-11-13 10:41:21.581665156 +0100
@@ -297,9 +297,9 @@ static int write_passwd(TFILE *fp,MYLDAP
   int numuids;
   char gidbuf[32];
   gid_t gid;
-  char gecos[100];
-  char homedir[100];
-  char shell[100];
+  char gecos[1024];
+  char homedir[256];
+  char shell[64];
   int i,j;
   /* get the usernames for this entry */
   usernames=myldap_get_values(entry,attmap_passwd_uid);
