diff -up ./nslcd/cfg.c.tls_crash ./nslcd/cfg.c
--- ./nslcd/cfg.c.tls_crash	2015-01-28 17:40:38.760756904 +0100
+++ ./nslcd/cfg.c	2015-01-28 17:45:34.608598561 +0100
@@ -423,12 +423,10 @@ static void get_restdup(const char *file
                         char **var)
 {
   check_argumentcount(filename,lnr,keyword,(*line!=NULL)&&(**line!='\0'));
-  if ((*var==NULL)||(strcmp(*var,*line)!=0))
-  {
-    /* Note: we have a memory leak here if a single mapping is changed
-             multiple times in one config (deemed not a problem) */
-    *var=xstrdup(*line);
-  }
+  /* Note: we have a memory leak here if a single mapping is changed
+           multiple times in one config (deemed not a problem) */
+  *var=xstrdup(*line);
+  /* mark that we are at the end of the line */
   *line=NULL;
 }
 
