diff -up nss-pam-ldapd-0.7.5/common/tio.c.fd_size nss-pam-ldapd-0.7.5/common/tio.c
--- nss-pam-ldapd-0.7.5/common/tio.c.fd_size	2013-02-19 09:12:36.167874336 +0100
+++ nss-pam-ldapd-0.7.5/common/tio.c	2013-02-19 09:14:39.915869687 +0100
@@ -184,6 +184,11 @@ static int tio_select(TFILE *fp, int rea
   while (1)
   {
     /* prepare our filedescriptorset */
+    if (fp->fd>=FD_SETSIZE)
+    {
+        errno=EBADFD;
+        return -1;
+    }
     FD_ZERO(&fdset);
     FD_SET(fp->fd,&fdset);
     /* figure out the time we need to wait */
@@ -433,6 +438,11 @@ static int tio_flush_nonblock(TFILE *fp)
   fd_set fdset;
   int rv;
   /* prepare our filedescriptorset */
+  if (fp->fd>=FD_SETSIZE)
+  {
+      errno=EBADFD;
+      return -1;
+  }
   FD_ZERO(&fdset);
   FD_SET(fp->fd,&fdset);
   /* set the timeout to 0 to poll */
