diff -up nss-pam-ldapd-0.8.14/nslcd/nslcd.c.retcode nss-pam-ldapd-0.8.14/nslcd/nslcd.c
--- nss-pam-ldapd-0.8.14/nslcd/nslcd.c.retcode	2017-02-08 09:52:39.687834074 +0100
+++ nss-pam-ldapd-0.8.14/nslcd/nslcd.c	2017-02-08 09:52:54.630891580 +0100
@@ -866,5 +866,5 @@ int main(int argc,char *argv[])
       log_log(LOG_ERR,"thread %d is still running, shutting down anyway",i);
   }
   /* we're done */
-  return EXIT_FAILURE;
+  return EXIT_SUCCESS;
 }
