diff -up nss-pam-ldapd-0.7.5/nss/common.h.erange nss-pam-ldapd-0.7.5/nss/common.h
--- nss-pam-ldapd-0.7.5/nss/common.h.erange	2015-11-12 10:07:52.922607974 +0100
+++ nss-pam-ldapd-0.7.5/nss/common.h	2015-11-12 10:13:42.395513718 +0100
@@ -84,11 +84,16 @@
   int32_t tmpint32; \
   enum nss_status retv; \
   /* check that we have a valid buffer */ \
-  if ((buffer==NULL)||(buflen<=0)) \
+  if ((buffer==NULL)||(buflen=0)) \
   { \
       *errnop=EINVAL; \
       return NSS_STATUS_UNAVAIL; \
   } \
+  if (buflen==0) \
+  { \
+    *errnop=ERANGE; \
+    return NSS_STATUS_TRYAGAIN; \
+  } \
   /* open socket and write request */ \
   NSLCD_REQUEST(fp,action,writefn); \
   /* read response */ \
@@ -133,7 +138,7 @@
   int32_t tmpint32; \
   enum nss_status retv; \
   /* check that we have a valid buffer */ \
-  if ((buffer==NULL)||(buflen<=0)) \
+  if ((buffer==NULL)||(buflen=0)) \
   { \
       /* close stream */ \
       if (fp!=NULL) \
@@ -145,6 +150,11 @@
       *errnop=EINVAL; \
       return NSS_STATUS_UNAVAIL; \
   } \
+  if (buflen==0) \
+  { \
+    *errnop=ERANGE; \
+    return NSS_STATUS_TRYAGAIN; \
+  } \
   /* check that we have a valid file descriptor */ \
   if (fp==NULL) \
   { \
