diff -up nss-pam-ldapd-0.8.13/nslcd/myldap.c.long_password nss-pam-ldapd-0.8.13/nslcd/myldap.c
--- nss-pam-ldapd-0.8.13/nslcd/myldap.c.long_password	2017-10-24 12:38:29.315411416 +0200
+++ nss-pam-ldapd-0.8.13/nslcd/myldap.c	2017-10-24 12:38:52.727517587 +0200
@@ -88,7 +88,7 @@ struct ldap_session
   /* the username to bind with */
   char binddn[256];
   /* the password to bind with if any */
-  char bindpw[64];
+  char bindpw[128];
   /* timestamp of last activity */
   time_t lastactivity;
   /* index into ldc_uris: currently connected LDAP uri */
diff -up nss-pam-ldapd-0.8.13/nslcd/pam.c.long_password nss-pam-ldapd-0.8.13/nslcd/pam.c
--- nss-pam-ldapd-0.8.13/nslcd/pam.c.long_password	2017-10-24 12:39:50.761780765 +0200
+++ nss-pam-ldapd-0.8.13/nslcd/pam.c	2017-10-24 12:41:15.083163153 +0200
@@ -246,7 +246,7 @@ int nslcd_pam_authc(TFILE *fp,MYLDAP_SES
   int rc;
   char username[256];
   char servicename[64];
-  char password[64];
+  char password[128];
   const char *userdn;
   MYLDAP_ENTRY *entry;
   int authzrc=NSLCD_PAM_SUCCESS;
@@ -617,8 +617,8 @@ int nslcd_pam_pwmod(TFILE *fp,MYLDAP_SES
   char userdn[256];
   int asroot;
   char servicename[64];
-  char oldpassword[64];
-  char newpassword[64];
+  char oldpassword[128];
+  char newpassword[128];
   const char *binddn=NULL; /* the user performing the modification */
   MYLDAP_ENTRY *entry;
   char authzmsg[1024];
